documentation/ImplementingASemanticCheck.md

```diff
@@ -0,0 +1,833 @@
+contained source location information since it had the field `CharBlock
+source`.  Thus, if I visited a `parser::Expr` node, I could get the source
+location information for the associated expression.
+
+## Determining the `INTENT`
+I knew that I could find the `INTENT` of the dummy argument associated with the
+actual argument from the function called `dummyIntent()` in the class
+`evaluate::ActualArgument` in the file `include/flang/evaluate/call.h`.  So
+if I could find an `evaluate::ActualArgument` in an expression, I could
+  determine the `INTENT` of the associated dummy argument.  I knew that it was
+  valid to call `dummyIntent()` because the data on which `dummyIntent()`
+  depends is established during semantic processing for expressions, and the
+  semantic processing for expressions happens before semantic checking for DO
+  constructs.
+
+In my prior work on checking the INTENT of arguments for SUBROUTINE calls,
+the parse tree held a node for the call (a `parser::CallStmt`) that contained
+an `evaluate::ProcedureRef` node.
+```C++
+  struct CallStmt {
+    WRAPPER_CLASS_BOILERPLATE(CallStmt, Call);
+    mutable std::unique_ptr<evaluate::ProcedureRef,
+        common::Deleter<evaluate::ProcedureRef>>
+        typedCall;  // filled by semantics
+  };
+```
+The `evaluate::ProcedureRef` contains a list of `evaluate::ActualArgument`
+nodes.  I could then find the INTENT of a dummy argument from the
+`evaluate::ActualArgument` node.
+
+For a FUNCTION call, though, there is no similar way to get from a parse tree
+node to an `evaluate::ProcedureRef` node.  But I knew that there was an
+existing framework used in DO construct semantic checking that traversed an
+`evaluate::Expr` node collecting `semantics::Symbol` nodes.  I guessed that I'd
+be able to use a similar framework to traverse an `evaluate::Expr`  node to
+find all of the `evaluate::ActualArgument` nodes.  
+
+Note that there are two distinct data types in the compiler called `Expr`.  One
+is in the `parser` namespace.  `parser::Expr` is defined in the file
+`include/flang/parser/parse-tree.h`.  It represents a parsed expression that
+maps directly to the source code and has fields that specify any operators in
+the expression, the operands, and the source position of the expression.
+
+The second `Expr` type is in the `evaluate` namespace.   The `evaluate`
+namespace contains many types associated with semantic checking of expressions.
+`evaluate::Expr` is defined in the file `include/flang/evaluate/expression.h`.
+It represents an expression after it has undergone semantic checking and
+contains information that is only available after semantic analysis.  This
+information includes the Fortran type of the expression, whether it's a
+reference to a function, whether it's an actual argument, etc.  After an
+expression has undergone semantic analysis, the field `typedExpr` in the
+`parser::Expr` node is filled in with a pointer to the analyzed expression in
+`evaluate::Expr`.
```

Comment:
Very nice, thanks!

Reply:
The `evaluate::Expr<T>` template classes, which is parameterized over the various types of Fortran, constitute a suite of strongly-typed representations of valid Fortran expressions of type `T` that has been fully elaborated with conversion operations and subjected to constant folding.